File: src/main/java/Connection.java
import java.io.DataInputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class Connection {

    private final Socket socket;

    public Connection(Socket socket) {
        this.socket = socket;
    }

    /**
     * Sends a message with correct encoding and so on
     */
    public void send(String message) throws IOException {
        // message must be utf-8 encoded
        byte[] encodedMessage = message.getBytes(StandardCharsets.UTF_8);

        // length cannot exceed signed int length, so max-unsigned-int is not checked atm
        long length = encodedMessage.length;

        // build unsigned int, as Java does not support unsigned variables
        byte[] uInt = new byte[4];
        uInt[0] = (byte) ((length & 0x00000000FF000000L) >> 24);
        uInt[1] = (byte) ((length & 0x0000000000FF0000L) >> 16);
        uInt[2] = (byte) ((length & 0x000000000000FF00L) >> 8);
        uInt[3] = (byte) ((length & 0x00000000000000FFL));

        OutputStream out = socket.getOutputStream();

        out.write(uInt);
        out.write(encodedMessage);
        out.flush();
    }

    /**
     * Receives a message with correct encoding and so on
     */
    public String receive() throws IOException {
        // DataInputStream has some convenience methods
        DataInputStream in = new DataInputStream(socket.getInputStream());

        // preamble is 4 bytes long
        int uInt = in.readInt();

        // convert to long for correct unsigned integer value
        long length = uInt & 0xffffffffL;

        // cannot allocate more than Integer.MAX_VALUE array space
        // this is only half of what a message can theoretically be
        if (length > Integer.MAX_VALUE) {
            // cannot handle this atm
            throw new IOException("Message is longer than " + Integer.MAX_VALUE + " bytes, which is Java array maximum.");
        }
        int intLength = (int) length;

        byte[] responseBytes = new byte[intLength];
        in.readFully(responseBytes);

        // message is utf-8 encoded
        return new String(responseBytes, StandardCharsets.UTF_8);
    }

    public void close() throws IOException {
        socket.close();
    }
}

File: src/main/scala/lib/Connection.scala
package lib

import java.io.{IOException, InputStream, OutputStream}
import java.net.Socket
import java.nio.charset.StandardCharsets
import java.nio.{ByteBuffer, ByteOrder}


case class Response(ack: Int) derives upickle.default.ReadWriter

class Connection(socket: Socket) {
  private val in = new InputStreamBuffer(socket.getInputStream)
  private val out = socket.getOutputStream

  def intToBytes(i: Int): Array[Byte] = ByteBuffer.allocate(Integer.BYTES).order(ByteOrder.BIG_ENDIAN).putInt(i).array()
  def bytesToInt(array: Array[Byte]): Int = ByteBuffer.wrap(array).order(ByteOrder.BIG_ENDIAN).getInt

  /** Sends a json control message
   *
   * @param message valid json object string
   * @throws IOException on any communication error
   */
  def sendControlMessage(message: String): Unit = {
    send(message)
    val responseString = receive()
    val response       = upickle.default.read[Response](responseString)
    if (response.ack != 1) throw new IOException("Ack failed. Return object is unexpected: " + response)
  }

  /** Sends a message with correct encoding and so on */
  def send(message: String): Unit = {
    // message must be utf-8 encoded
    val encodedMessage = message.getBytes(StandardCharsets.UTF_8)
    // length cannot exceed signed int length, so max-unsigned-int is not checked atm
    out.write(intToBytes(encodedMessage.length))
    out.write(encodedMessage)
    out.flush()
  }

  /** Receives a message with correct encoding and so on */
  def receive(): String = {
    // preamble is 4 bytes long
    in.readNBytes(4)
    val length = bytesToInt(in.buffer.slice(0, 4))
    // cannot allocate more than Integer.MAX_VALUE array space
    // this is only half of what a message can theoretically be
    if (Integer.compareUnsigned(length, Integer.MAX_VALUE) > 0) { // cannot handle this atm
      throw new IOException("Message is longer than " + Integer.MAX_VALUE + " bytes, which is Java array maximum.")
    }
    // message is utf-8 encoded
    in.readNBytes(length)
    new String(in.buffer.slice(0, length), StandardCharsets.UTF_8)
  }

  def close(): Unit = socket.close()
}

class InputStreamBuffer(val in: InputStream) {
  var buffer = new Array[Byte](1)

  def readNBytes(length: Int): Unit = {
    if (buffer.length < length) {
      buffer = new Array[Byte](length)
    }

    var totalBytesRead = 0
    var currentBytesRead = 0
    while (totalBytesRead < length) {
      currentBytesRead = in.read(buffer, totalBytesRead, length-totalBytesRead)

      if (currentBytesRead == -1) {
        throw new Exception("socket closed")
      }

      totalBytesRead = totalBytesRead + currentBytesRead
    }
  }
}

